Use system functions to allocate aligned memory

Lastique opened the pull request:
### Description 

This uses the standard allocator for allocating and freeing memory with higher alignment requirements instead of the hand-rolled implementation.

- [x] - git commit message contains appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/oneapi-src/oneTBB/blob/master/CONTRIBUTING.md#pull-requests) for details)_

### Type of change

- [ ] bug fix - _change which fixes an issue_
- [ ] new feature - _change which adds functionality_
- [ ] tests - _change in tests_
- [ ] infrastructure - _change in infrastructure and CI_
- [ ] documentation - _documentation update_
- [x] optimization

### Tests

- [ ] added - _required for new features and for some bug fixes_
- [x] not needed

### Documentation

- [ ] updated in # - _add PR number_
- [ ] needs to be updated
- [x] not needed

### Breaks backward compatibility
- [ ] Yes
- [x] No
- [ ] Unknown

Lastique commented:
This patch originated from https://github.com/oneapi-src/oneTBB/pull/326.

alexey-katranov commented on src/tbb/allocator.cpp:
```diff
@@ -67,11 +85,15 @@ static void  (*deallocate_handler)(void* pointer) = nullptr;
 //! Initialization routine used for first indirect call via cache_aligned_allocate_handler.
 static void* initialize_cache_aligned_allocate_handler(std::size_t n, std::size_t alignment);
 
+#if !defined(__TBB_USE_MSVC_ALIGNED_MALLOC)
 //! Allocates memory using standard malloc. It is used when scalable_allocator is not available
 static void* std_cache_aligned_allocate(std::size_t n, std::size_t alignment);
+#endif
 
+#if !defined(__TBB_USE_POSIX_MEMALIGN) && !defined(__TBB_USE_MEMALIGN) && !defined(__TBB_USE_MSVC_ALIGNED_MALLOC)
 //! Allocates memory using standard free. It is used when scalable_allocator is not available
 static void  std_cache_aligned_deallocate(void* p);
+#endif
```
What about the approach like in `std_cache_aligned_allocate`?
```suggestion
#if !defined(__TBB_USE_MSVC_ALIGNED_MALLOC)
//! Allocates memory using standard malloc. It is used when scalable_allocator is not available
static void* std_cache_aligned_allocate(std::size_t n, std::size_t alignment);

#if !defined(__TBB_USE_POSIX_MEMALIGN) && !defined(__TBB_USE_MEMALIGN)
//! Allocates memory using standard free. It is used when scalable_allocator is not available
static void  std_cache_aligned_deallocate(void* p);
#endif
#endif // !defined(__TBB_USE_MSVC_ALIGNED_MALLOC)
```

Lastique replied:
Done.

alexey-katranov commented on src/tbb/allocator.cpp:
```diff
@@ -67,11 +85,15 @@ static void  (*deallocate_handler)(void* pointer) = nullptr;
 //! Initialization routine used for first indirect call via cache_aligned_allocate_handler.
 static void* initialize_cache_aligned_allocate_handler(std::size_t n, std::size_t alignment);
 
+#if !defined(__TBB_USE_MSVC_ALIGNED_MALLOC)
 //! Allocates memory using standard malloc. It is used when scalable_allocator is not available
 static void* std_cache_aligned_allocate(std::size_t n, std::size_t alignment);
 
+#if !defined(__TBB_USE_POSIX_MEMALIGN) && !defined(__TBB_USE_MEMALIGN)
 //! Allocates memory using standard free. It is used when scalable_allocator is not available
 static void  std_cache_aligned_deallocate(void* p);
+#endif // !defined(__TBB_USE_POSIX_MEMALIGN) && !defined(__TBB_USE_MEMALIGN)
+#endif // !defined(__TBB_USE_MSVC_ALIGNED_MALLOC)
```
```suggestion
//! Allocates memory using standard malloc. It is used when scalable_allocator is not available
static void* std_cache_aligned_allocate(std::size_t n, std::size_t alignment);

//! Allocates memory using standard free. It is used when scalable_allocator is not available
static void  std_cache_aligned_deallocate(void* p);
```

alexey-katranov commented on src/tbb/allocator.cpp:
```diff
@@ -123,8 +145,16 @@ void initialize_handler_pointers() {
         // which forces them to wait.
         allocate_handler_unsafe = &std::malloc;
         deallocate_handler = &std::free;
+#if defined(__TBB_USE_POSIX_MEMALIGN) || defined(__TBB_USE_MEMALIGN)
+        cache_aligned_allocate_handler_unsafe = &std_cache_aligned_allocate;
+        cache_aligned_deallocate_handler = &::free;
+#elif defined(__TBB_USE_MSVC_ALIGNED_MALLOC)
+        cache_aligned_allocate_handler_unsafe = &::_aligned_malloc;
+        cache_aligned_deallocate_handler = &::_aligned_free;
+#else
         cache_aligned_allocate_handler_unsafe = &std_cache_aligned_allocate;
         cache_aligned_deallocate_handler = &std_cache_aligned_deallocate;
+#endif
```
```suggestion
        cache_aligned_allocate_handler_unsafe = &std_cache_aligned_allocate;
        cache_aligned_deallocate_handler = &std_cache_aligned_deallocate;
```

alexey-katranov commented on src/tbb/allocator.cpp:
```diff
@@ -184,7 +214,17 @@ void __TBB_EXPORTED_FUNC cache_aligned_deallocate(void* p) {
+    if (res != 0)
+        p = nullptr;
+    return p;
+#else
```
```suggestion
#elif defined(__TBB_USE_MSVC_ALIGNED_MALLOC)
    return _aligned_malloc(bytes, alignment);
#else
```

alexey-katranov commented on src/tbb/allocator.cpp:
```diff
@@ -184,7 +214,17 @@ void __TBB_EXPORTED_FUNC cache_aligned_deallocate(void* p) {
     (*cache_aligned_deallocate_handler)(p);
 }
 
+#if !defined(__TBB_USE_MSVC_ALIGNED_MALLOC)
```
```suggestion
```

alexey-katranov commented on src/tbb/allocator.cpp:
```diff
@@ -199,8 +239,11 @@ static void* std_cache_aligned_allocate(std::size_t bytes, std::size_t alignment
     return reinterpret_cast<void*>(result);
+#endif
 }
+#endif // !defined(__TBB_USE_MSVC_ALIGNED_MALLOC)
```
```suggestion
```

alexey-katranov commented on src/tbb/allocator.cpp:
```diff
@@ -199,8 +239,11 @@ static void* std_cache_aligned_allocate(std::size_t bytes, std::size_t alignment
 }
+#endif // !defined(__TBB_USE_MSVC_ALIGNED_MALLOC)
 
+#if !defined(__TBB_USE_POSIX_MEMALIGN) && !defined(__TBB_USE_MEMALIGN) && !defined(__TBB_USE_MSVC_ALIGNED_MALLOC)
 static void std_cache_aligned_deallocate(void* p) {
```
```suggestion
static void std_cache_aligned_deallocate(void* p) {
#if defined(__TBB_USE_POSIX_MEMALIGN) || defined(__TBB_USE_MEMALIGN)
    free(p);
#elif defined(__TBB_USE_MSVC_ALIGNED_MALLOC)
    _aligned_free(p);
#else
```

alexey-katranov commented on src/tbb/allocator.cpp:
```diff
@@ -210,6 +253,7 @@ static void std_cache_aligned_deallocate(void* p) {
         std::free(reinterpret_cast<void*>(base));
     }
 }
+#endif // !defined(__TBB_USE_POSIX_MEMALIGN) && !defined(__TBB_USE_MEMALIGN) && !defined(__TBB_USE_MSVC_ALIGNED_MALLOC)
```
```suggestion
#endif // defined(__TBB_USE_POSIX_MEMALIGN) || defined(__TBB_USE_MEMALIGN)
}
```

alexey-katranov reviewed (commented):
Overall, I am Ok with your approach. However, can you consider a set of proposed changes to reduce the code slicing and make macro dependent code more isolated inside the functions? What do you think? (I understand that it will cause one more call in the chain but I hope modern compilers can perform tail call optimization)

Lastique commented:
> Overall, I am Ok with your approach. However, can you consider a set of proposed changes to reduce the code slicing and make macro dependent code more isolated inside the functions? What do you think? (I understand that it will cause one more call in the chain but I hope modern compilers can perform tail call optimization)

I'm quite certain the compiler won't optimize away the forwarding function, even if it amounts to a `jmp` instruction, because it has to have a unique address. That's unnecessary overhead that I'd prefer to avoid. The resulting code doesn't improve much in terms if simplicity, IMHO, and I value runtime performance more than code simplicity anyway. However, if you insist I can change the code as you suggest. Please, confirm.

alexey-katranov commented:
> The resulting code doesn't improve much in terms if simplicity

At rough estimation, I removed 6 condition blocks but added only 2.

> I value runtime performance more than code simplicity anyway.

It seems we have slightly different views here. The code is not on critical path in term of performance but any correctness issue should be avoided. However, code complexity wastes engineering time for debug and maintaining while it is not critical for performance of the whole library and we can spend this time to improve something else. As for optimization, it seems possible only for `__TBB_USE_MSVC_ALIGNED_MALLOC`, while other platforms will have excessive `jmp`. So, why not have a generic approach for all platforms until we face performance issue due to this inefficiency?

>  However, if you insist I can change the code as you suggest. Please, confirm.

I do not like to insist and prefer finding consensus. In any case, good balance is consist of multiple opinions. Taking into consideration that you contribute relatively often and reasonably, your opinion is valuable.

Lastique commented:
> > The resulting code doesn't improve much in terms if simplicity
> 
> At rough estimation, I removed 6 condition blocks but added only 2.

Yes, it is better, I agree, but I just don't find the change *that much* better. I mean, the original code is not hard to decipher, and all conditions are uniform and easy to locate if they need to be updated. In my personal code I would probably be fine with the extra preprocessor checks if it makes the code more optimal.

> > I value runtime performance more than code simplicity anyway.
> 
> It seems we have slightly different views here. The code is not on critical path in term of performance but any correctness issue should be avoided. However, code complexity wastes engineering time for debug and maintaining while it is not critical for performance of the whole library and we can spend this time to improve something else. As for optimization, it seems possible only for `__TBB_USE_MSVC_ALIGNED_MALLOC`, while other platforms will have excessive `jmp`. So, why not have a generic approach for all platforms until we face performance issue due to this inefficiency?

Because MSVC is lucky enough for this optimization to be possible, and it seems wasteful to not take this opportunity.

> > However, if you insist I can change the code as you suggest. Please, confirm.
> 
> I do not like to insist and prefer finding consensus. In any case, good balance is consist of multiple opinions. Taking into consideration that you contribute relatively often and reasonably, your opinion is valuable.

The thing is, this project is not my personal code, and you, the maintainers, may have different priorities than I do, and this is totally fine. I'm just stating my position, and if you still think you prefer it the other way, I can change accordingly, and that is also fine. If I'm adamantly against something, you can be sure I will say so and refuse to change. :)

In fact in this case I don't care about MSVC as my main target platform is Linux. I just added support for it because it was the straightforward and easy thing to do. My main goal here is to get this optimization merged, primarily for Linux, and I don't want this PR to hang indefinitely in limbo or get bogged down in discussions about minor things like this. So I will change the code as you suggest, no problem.

Lastique commented:
> Because MSVC is lucky enough for this optimization to be possible, and it seems wasteful to not take this opportunity.

Actually, it's not only MSVC - `free` is used directly on other platforms.

Lastique commented:
> So I will change the code as you suggest, no problem.

Done.

alexey-katranov reviewed (approved):
Thank you for the contribution.
